quint/src/sourceResolver.ts

```diff
@@ -0,0 +1,166 @@
+/* ----------------------------------------------------------------------------------
+ * Copyright (c) Informal Systems 2023. All rights reserved.
+ * Licensed under the Apache 2.0.
+ * See License.txt in the project root for license information.
+ * --------------------------------------------------------------------------------- */
+
+/**
+ * Source resolution for Quint. This module deals with loading sources from strings
+ * and files.
+ *
+ * @author Igor Konnov
+ *
+ * @module
+ */
+import { Either, left, right } from '@sweet-monads/either'
+import { dirname } from 'path'
+import { readFileSync } from 'fs'
+import { posix } from 'path'
+import { lf } from 'eol'
+
+/**
+ * An abstraction of a full path that is used to distinguish the same module,
+ * when it is resolved via different paths. Note that this path may be
+ * system-dependent and resolver dependent.
+ */
+export interface SourceLookupPath {
+  /**
+   * An internal representation of a path, which may be system-dependent.
+   * The important property is that this path is normalized, that is,
+   * when the same module is referred via different paths, we should still
+   * arrive at the same internal path representation.
+   */
+  normalizedPath: string
+
+  /**
+   * Produce the source name that will be shown in errors and stored in source maps.
+   * @returns a string representation of the path that is written in source maps
+   */
+  toSourceName: () => string
+}
+
+/**
+ * A general interface for resolving sources.
+ */
+export interface SourceResolver {
```

Comment:
I've introduced this abstraction, as it has been quite an issue in SANY. We do not have to deal with filename, in order to resolve relative paths. In case of tests, we could use a map of strings. For a web interface, we could use a map of names to URIs. It's true that the main use case is filenames right now.

Reply:
I see. That makes sense. Thanks.

Reply:
Actually, I just have hit this problem with the VSCode plugin. @bugarela and I agreed to handle only `file://` atm, but we should check how it is done with URIs in general, postponed for #831.